README.md
Name: Arthur Lu
PID: A16335944
Site Link: https://monumental-crisp-e70991.netlify.app/

changelog.md
# Style Changes
- improve page style and layout
	- use a new nav bar layout that occupies less space while providing more navigational links
	- use a card style content layout for some content on the site
	- added placeholder content including mission statement
# Interaction Changes
- use Octokit (3rd party script) and the github api to fetch repositories automatically
	- allows the site to be populated with projects automatically

scripts/blog.js
import {createDialog} from "./customdialogs.js";

// local storage items:
// nextID: next id to use for new post
// posts: array of post objects

window.addEventListener("DOMContentLoaded", init);

function init () {
	if (!getAllPosts()) {
		localStorage.setItem("nextPostID", 3);
		saveAllPosts([
			{
				id: 0,
				title: "My First Blog Post",
				date: "2023/2/24",
				summary: "A sample first post"
			},
			{
				id: 1,
				title: "My Second Blog Post",
				date: "2023/2/25",
				summary: "A sample second post"
			},
			{
				id: 2,
				title: "My Third Blog Post",
				date: "2023/2/26",
				summary: "A sample third post"
			}
		]);
	}
	populatePosts();
	document.querySelector("#add-post").addEventListener("click", handleAddPost);
}

function populatePosts () {
	let container = document.querySelector("#posts");
	container.innerHTML = "";
	let posts = getAllPosts();
	posts.forEach((element) => {
		let post = document.createElement("li");
		post.id = element.id;
		post.innerHTML = `<p>Post: ${element.title}</p><p>Date: ${element.date}</p><p>Summary: ${element.summary}</p><button class="edit">Edit</button><button class="delete">Delete</button>`;
		post.querySelector(".edit").addEventListener("click", handleEditPost.bind(post));
		post.querySelector(".delete").addEventListener("click", handleDeletePost.bind(post));
		container.append(post)
	});
}

function addPost (post) {
	post.id = getNextID();
	let posts = getAllPosts();
	posts.push(post);
	saveAllPosts(posts);
}

function getPost (id) {
	let posts = getAllPosts();
	posts.forEach((element) => {
		if (element.id === id) {
			return element;
		}
	});
	return null;
}

function getAllPosts () {
	return JSON.parse(localStorage.getItem("posts"));
}

function saveAllPosts (posts) {
	localStorage.setItem("posts", JSON.stringify(posts));
}

function updatePost (id, post) {
	post.id = id;
	let index = null;
	let posts = getAllPosts();
	for (let i = 0; i < posts.length; i++) {
		if (posts[i].id === id) {
			index = i;
			break;
		}
	}
	if (index != null) {
		posts[index] = post;
		saveAllPosts(posts);
		return true;
	}
	return false;
}

function deletePost (id) {
	let index = null;
	let posts = getAllPosts();
	for (let i = 0; i < posts.length; i++) {
		if (posts[i].id === id) {
			index = i;
			break;
		}
	}
	if (index != null) {
		posts.splice(index, 1);
		saveAllPosts(posts);
		return true;
	}
	return false;
}

function getNextID () {
	let nextID = parseInt(localStorage.getItem("nextPostID"));
	localStorage.setItem("nextPostID", nextID+1);
	return nextID;
}

function handleAddPost () {
	createDialog("#blog-template", "Create A Post", (result, form) => {
		if (result) {
			addPost({
				title: form.get("title"),
				date: form.get("date"),
				summary: form.get("summary")
			});
			populatePosts();
		}
	});
}

function handleEditPost () { // this bound to element, this.id should be clicked post id
	createDialog("#blog-template", "Edit Post", (result, form) => {
		if (result) {
			updatePost(parseInt(this.id), {
				title: form.get("title"),
				date: form.get("date"),
				summary: form.get("summary")
			});
			populatePosts();
		}
	})
}

function handleDeletePost () {
	createDialog("#confirm-template", "Delete Post", (result, form) => {
		if (result) {
			deletePost(parseInt(this.id));
			populatePosts();
		}
	})
}

scripts/projects.js
import { Octokit } from "https://cdn.skypack.dev/@octokit/core";

window.addEventListener("DOMContentLoaded", init);

async function init () {
	const octokit = new Octokit()

	let repos = await octokit.request('GET /users/ltcptgeneral/repos', {
		username: 'ltcptgeneral'
	});

	let container = document.querySelector("#projects");

	repos.data.forEach((element) => {
		let content = document.querySelector("#project-article").content.cloneNode(true);
		content.querySelector("#project-name").innerText = element.name;
		content.querySelector("#project-desc").innerText = element.description;
		content.querySelector("#project-lang").innerText = element.language;
		content.querySelector("article").addEventListener("click", () => {
			window.location.assign(element.html_url);
		});
		container.append(content)
	});
}

scripts/customdialogs.js
/* createDialog creates a new dialog on the page given a template with content, and an optional callback function
 *			NOTE: the template is assumed to have at least a content tag with id message
 *
 * template: query string corresponding to the template element used as the basis for the dialog
 * message: message to display with the dialog
 * callback: callback function (result, form) called when dialog is closed and 2 parameters:
 * 			result: whether the form was submitted or canceled as true/false
 * 			form: string containing the user input
 * 			NOTE: the callback should not change the dialog or form DOM nodes in any way!!!
 */
export function createDialog (template, message, callback=()=>{}) {
	let content = document.querySelector(template).content.cloneNode(true);

	let form = document.createElement("form");
	form.method = "dialog";
	form.append(content);

	let dialog = document.createElement("dialog");
	dialog.append(form);

	dialog.querySelector("#message").innerText = message;
	if (callback) {
		dialog.addEventListener("close", () => {
			callback(dialog.returnValue === "OK", new FormData(form));
			dialog.remove();
		});
	}
	document.body.append(dialog);
	dialog.showModal();
}

scripts/styledblog.js
import {createDialog} from "./customdialogs.js";

// local storage items:
// nextID: next id to use for new post
// posts: array of post objects

window.addEventListener("DOMContentLoaded", init);

function init () {
	if (!getAllPosts()) {
		localStorage.setItem("nextPostID", 3);
		saveAllPosts([
			{
				id: 0,
				title: "My First Blog Post",
				date: "2023/2/24",
				summary: "A sample first post"
			},
			{
				id: 1,
				title: "My Second Blog Post",
				date: "2023/2/25",
				summary: "A sample second post"
			},
			{
				id: 2,
				title: "My Third Blog Post",
				date: "2023/2/26",
				summary: "A sample third post"
			}
		]);
	}
	populatePosts();
	document.querySelector("#add-post").addEventListener("click", handleAddPost);
}

function populatePosts () {
	let container = document.querySelector("#posts");
	container.innerHTML = "";
	let posts = getAllPosts();
	posts.forEach((element) => {
		let post = document.createElement("li");
		post.id = element.id;
		post.innerHTML = `
			<h2>Post: ${element.title}</h2>
			<date>Date: ${element.date}</date>
			<p>Summary: ${element.summary}</p>
			<div class="btn-group">
				<img class="edit" src="images/edit.svg" width="32px">
				<img class="delete" src="images/delete.svg" width="32px">
			</div>	
		`;
		post.querySelector(".edit").addEventListener("click", handleEditPost.bind(post));
		post.querySelector(".delete").addEventListener("click", handleDeletePost.bind(post));
		container.append(post)
	});
}

function addPost (post) {
	post.id = getNextID();
	let posts = getAllPosts();
	posts.push(post);
	saveAllPosts(posts);
}

function getPost (id) {
	let posts = getAllPosts();
	posts.forEach((element) => {
		if (element.id === id) {
			return element;
		}
	});
	return null;
}

function getAllPosts () {
	return JSON.parse(localStorage.getItem("posts"));
}

function saveAllPosts (posts) {
	localStorage.setItem("posts", JSON.stringify(posts));
}

function updatePost (id, post) {
	post.id = id;
	let index = null;
	let posts = getAllPosts();
	for (let i = 0; i < posts.length; i++) {
		if (posts[i].id === id) {
			index = i;
			break;
		}
	}
	if (index != null) {
		posts[index] = post;
		saveAllPosts(posts);
		return true;
	}
	return false;
}

function deletePost (id) {
	let index = null;
	let posts = getAllPosts();
	for (let i = 0; i < posts.length; i++) {
		if (posts[i].id === id) {
			index = i;
			break;
		}
	}
	if (index != null) {
		posts.splice(index, 1);
		saveAllPosts(posts);
		return true;
	}
	return false;
}

function getNextID () {
	let nextID = parseInt(localStorage.getItem("nextPostID"));
	localStorage.setItem("nextPostID", nextID+1);
	return nextID;
}

function handleAddPost () {
	createDialog("#blog-template", "Create A Post", (result, form) => {
		if (result) {
			addPost({
				title: form.get("title"),
				date: form.get("date"),
				summary: form.get("summary")
			});
			populatePosts();
		}
	});
}

function handleEditPost () { // this bound to element, this.id should be clicked post id
	createDialog("#blog-template", "Edit Post", (result, form) => {
		if (result) {
			updatePost(parseInt(this.id), {
				title: form.get("title"),
				date: form.get("date"),
				summary: form.get("summary")
			});
			populatePosts();
		}
	})
}

function handleDeletePost () {
	createDialog("#confirm-template", "Delete Post", (result, form) => {
		if (result) {
			deletePost(parseInt(this.id));
			populatePosts();
		}
	})
}